widget/slider.go

```diff
@@ -0,0 +1,289 @@
+	s.show(s)
+}
+
+// Hide this widget, if it was previously visible
+func (s *Slider) Hide() {
+	s.hide(s)
+}
+
+// Move the widget to a new position, relative to its parent.
+func (s *Slider) Move(pos fyne.Position) {
+	s.move(pos, s)
+}
+
+// DragEnd function.
+func (s *Slider) DragEnd() {
+}
+
+// Dragged function.
+func (s *Slider) Dragged(e *fyne.DragEvent) {
+	ok, pos, max := s.wasSliderEvent(&(e.PointEvent))
+
+	if ok {
+		// clamp the position for drags that go out of bounds
+		if pos > max {
+			pos = max
+		} else if pos < 0 {
+			pos = 0
+		}
+
+		s.fireTrigger(pos, max)
+	}
+}
+
+func (s *Slider) wasSliderEvent(e *fyne.PointEvent) (
+	ok bool, pos, max int) {
+
+	render := Renderer(s).(*sliderRenderer)
+
+	tp := render.thumb.Position()
+	ts := render.thumb.Size()
+
+	pad := theme.Padding()
+
+	x := e.Position.X
+	y := e.Position.Y
+
+	if s.Vertical {
+		if x > (tp.X-pad) && x < (tp.X+ts.Width+pad) {
+			// if the cursor was inside the slider area
+			return true, y, render.track.Size().Height
+		}
+	} else {
+		if y > (tp.Y-pad) && y < (tp.Y+ts.Height+pad) {
+			// if the cursor was inside the slider area
+			return true, x, render.track.Size().Width
+		}
+	}
+	return false, 0, 0
+}
+
+func (s *Slider) fireTrigger(pos, max int) {
+	// update value
+	s.updateValue(float64(pos) / float64(max))
+	Refresh(s)
+
+	if s.OnChanged != nil {
+		s.OnChanged()
+	}
+}
+
+func (s *Slider) updateValue(ratio float64) {
+	if s.Vertical {
+		ratio = 1 - ratio
+	}
+	v := s.Min + ratio*(s.Max-s.Min)
+
+	i := -(math.Log10(s.Step))
+	p := math.Pow(10, i)
+
+	// hack to deal with asymptotic effect for decimal increments
+	if s.Step < 1 && math.Abs(v) == math.Abs(s.Max) {
+		s.Value = float64(int(math.Ceil(v*p)) / int(p))
+	} else {
+		s.Value = float64(int(v*p)) / p
+	}
+}
+
+// CreateRenderer is a private method to Fyne which links
+// this widget to its renderer
+func (s *Slider) CreateRenderer() fyne.WidgetRenderer {
+	track := canvas.NewRectangle(theme.ButtonColor())
+	active := canvas.NewRectangle(theme.TextColor())
+	thumb := &canvas.Circle{
+		FillColor:   theme.TextColor(),
+		StrokeWidth: 0}
+
+	objects := []fyne.CanvasObject{track, active, thumb}
+
+	return &sliderRenderer{track, active, thumb, objects, s}
+}
+
+type sliderRenderer struct {
+	track  *canvas.Rectangle
+	active *canvas.Rectangle
+	thumb  *canvas.Circle
+
+	objects []fyne.CanvasObject
+	slider  *Slider
+}
+
+// ApplyTheme is called when the Slider may need to update its look
+func (s *sliderRenderer) ApplyTheme() {
+	s.track.FillColor = theme.ButtonColor()
+	s.thumb.FillColor = theme.TextColor()
+	s.active.FillColor = theme.TextColor()
+	s.Refresh()
+}
+
+// Refresh is used to update the widget state for drawing
+func (s *sliderRenderer) Refresh() {
+	s.Layout(s.slider.Size())
+	canvas.Refresh(s.slider)
+}
+
+// Layout the components of the slider widget
+func (s *sliderRenderer) Layout(size fyne.Size) {
+	padLen := theme.Padding()
+	sideLen := padLen * standardScale
```

Comment:
Isn't this a diameter rather than a length?
Also the padding variable above doesn't seem to be a length (assuming that's what "Len" is short for?)

Reply:
> Isn't this a diameter rather than a length?

Semantically, isn't a diameter a length?

I am pretty sure they can be the same thing, though I am not sure that this shape is possible.😁 
![c](https://user-images.githubusercontent.com/20728117/63106296-24f38380-bf48-11e9-9b67-f314f3e5a625.png)

Semantics aside, I do believe you are referring to the naming convention...

Reply:
Whilst of course you are right I would be surprised if any APIs or popular code use "length" in the context of a circle's diameter... we should use obvious naming where possible.